File: python/canvastekk_workflow_sdk/request.py
"""
Node Execution Request Model

This is the payload sent to a node's /execute endpoint.
"""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field, model_validator


class NodeExecutionRequest(BaseModel):
    """
    Request payload for node execution.

    Sent by the orchestrator to a node's POST /execute endpoint.
    ``run_id``/``node_id`` are constrained to a slug charset and may not
    contain ``..`` (defense against path construction attacks — the ids
    flow into ``/tmp/{run_id}/{node_id}`` output paths).
    """

    _SLUG_RE = r"^[A-Za-z0-9._-]+$"

    run_id: str = Field(
        pattern=_SLUG_RE,
        description="Workflow run identifier (slug: letters, digits, dot, underscore, hyphen)",
    )
    node_id: str = Field(
        pattern=_SLUG_RE,
        description="Node instance ID in workflow (slug: letters, digits, dot, underscore, hyphen)",
    )

    @model_validator(mode="after")
    def _reject_dot_segments(self) -> NodeExecutionRequest:
        """Reject `..` substrings and dot-only values.

        The slug charset alone still permits `..` and `.` segments; these
        flow into ``/tmp/{run_id}/{node_id}`` path construction and would
        escape the run sandbox (DA-1711 3.1).
        """
        for field_name in ("run_id", "node_id"):
            value = getattr(self, field_name)
            if ".." in value or value.strip(".") != value or not value.strip("."):
                raise ValueError(
                    f"{field_name} must not contain dot segments (got {value!r})"
                )
        return self
    inputs: dict[str, Any] = Field(
        default_factory=dict,
        description="Input values (may include signed URLs for file access)",
    )
    callback_url: str | None = Field(
        default=None,
        description="For async execution - URL to POST result to when complete",
    )
    output_upload_url: dict[str, str] | None = Field(
        default=None,
        description="Mapping of output field name to pre-signed S3 PUT URL for uploading output files",
    )

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "run_id": "run-abc123",
                    "node_id": "echo-1",
                    "inputs": {"message": "Hello, World!"},
                    "callback_url": None,
                    "output_upload_url": None,
                }
            ]
        }
    }

File: typescript/src/request.ts
import { z } from "zod";

/**
 * Slug charset for run_id/node_id — letters, digits, dot, underscore,
 * hyphen. NOTE: the charset alone still permits `..` and `.` segments;
 * the refinement rejects them (the ids flow into filesystem paths).
 */
const slugField = z
  .string()
  .regex(/^[A-Za-z0-9._-]+$/, "must contain only letters, digits, '.', '_', '-'")
  .refine((v) => !v.includes("..") && !/^\.$/.test(v) && !/^\.\.+$/.test(v), {
    message: "must not contain dot segments",
  });

export const NodeExecutionRequestSchema = z.object({
  run_id: slugField,
  node_id: slugField,
  inputs: z.record(z.unknown()).default({}),
  callback_url: z.string().nullable().optional(),
  output_upload_url: z
    .union([z.record(z.string()), z.null()])
    .optional()
    .default(null),
});

/**
 * Node execution request payload sent by the workflow engine.
 *
 * Fields use snake_case for wire-format compatibility with the Python engine.
 *
 * @property run_id - Unique identifier for the workflow run
 * @property node_id - Unique identifier for this node within the run
 * @property inputs - Node input values (may include presigned URLs for file fields)
 * @property callback_url - Optional URL to post completion status
 * @property output_upload_url - Optional presigned URL(s) for uploading output files
 */
export type NodeExecutionRequest = z.infer<typeof NodeExecutionRequestSchema>;
